brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -140,101 +140,109 @@ public static TraceContextOrSamplingFlags parseB3SingleFormat(CharSequence b3) {
+    // they need to consider, and if there are not enough or too many fields.
+    int hyphenCount = 0;
+    int indexOfFirstHyphen = -1;
     for (int i = beginIndex; i < endIndex; i++) {
-      if (b3.charAt(i) >= 128) {
-        Platform.get().log("Invalid input: non-ASCII character at offset {0}", i, null);
+      char c = value.charAt(i);
+      if (c == '-') {
+        if (indexOfFirstHyphen == -1) {
+          indexOfFirstHyphen = i;
+        }
+        hyphenCount++;
+      } else if ((c < '0' || c > '9') && (c < 'a' || c > 'f')) {
+        Platform.get().log("Invalid input: only valid characters are lower-hex and hyphen", null);
         return null;
       }
     }
 
-    long traceIdHigh, traceId;
-    if (b3.charAt(pos + 32) == '-') {
-      traceIdHigh = tryParse16HexCharacters(b3, pos, endIndex);
-      pos += 16; // upper 64 bits of the trace ID
-      traceId = tryParse16HexCharacters(b3, pos, endIndex);
-    } else {
-      traceIdHigh = 0L;
-      traceId = tryParse16HexCharacters(b3, pos, endIndex);
-    }
-    pos += 16; // traceId
-
-    if (traceIdHigh == 0L && traceId == 0L) {
-      Platform.get().log("Invalid input: expected a 16 or 32 lower hex trace ID at offset 0", null);
+    if (hyphenCount < 1) {
+      Platform.get().log("Truncated reading trace ID", null);
       return null;
-    }
-
-    if (isLowerHex(b3.charAt(pos))) {
-      Platform.get().log("Invalid input: trace ID is too long", null);
+    } else if (hyphenCount > 3) {
+      Platform.get().log("Invalid input: more than 4 fields exist", null);
       return null;
     }
 
-    if (!checkHyphen(b3, pos++)) return null;
+    int pos = beginIndex;
 
-    long spanId = tryParse16HexCharacters(b3, pos, endIndex);
-    if (spanId == 0L) {
-      Platform.get().log("Invalid input: expected a 16 lower hex span ID at offset {0}", pos, null);
+    long traceIdHigh, traceId;
+    int traceIdLength = indexOfFirstHyphen - beginIndex;
+    if (traceIdLength == 32) {
+      traceIdHigh = lenientLowerHexToUnsignedLong(value, pos, pos + 16);
+      pos += 16; // upper 64 bits of the trace ID
+    } else if (traceIdLength == 16) {
+      traceIdHigh = 0L;
+    } else {
+      Platform.get().log("Invalid input: expected a 16 or 32 lower hex trace ID", null);
       return null;
     }
+
+    traceId = tryParseId("trace", value, pos, endIndex);
+    if (traceId == 0) return null;
+    pos += 17; // lower 64 bits of the trace ID and the hyphen
+
+    long spanId = tryParseId("span", value, pos, endIndex);
+    if (spanId == 0) return null;
     pos += 16; // spanid
 
     int flags = 0;
     long parentId = 0L;
     if (endIndex > pos) {
-      if (isLowerHex(b3.charAt(pos))) {
-        Platform.get().log("Invalid input: span ID is too long", null);
+      pos++; // consume the delimiter
+
+      if (endIndex == pos) { // traceid-spanid-
+        Platform.get().log("Truncated after reading span ID", null);
```

Comment:
If the precision in the error messages isn't important, it might be nice to have these checks in the validation loop above so it's all together.

Reply:
think we're good here.. only a small loosening made a lot of things possible. good idea